### ISSUE TITLE
[MU4 Task] Dropdowns, menus and lists should highlight items in reaction to key presses

### ISSUE THREAD

**Issue by cbjeukendrup** [musescore/MuseScore#16508]
**Task description**
When a dropdown, menu or list is focused, you should be able to type text on the keyboard, to quickly jump to the corresponding item. 

For example, in the dropdown shown below, 
- when you press `S`, it should highlight the first item that comes lexicographically after "S", so in this case "Sans Serif Collection".

- when you press `S` `E` in quick succession, it should highlight the first item that comes lexicographically after "Se", so in this case "Segoe Fluent Icons". 

<img src="https://media.discordapp.net/attachments/818848561047273544/1078352395236544623/image.png" width="160"/>

**Technical note**
This will need to be implemented for all QML dropdowns/menus/lists. 
In Qt Widgets, this already works out of the box (and for those few cases where it doesn't, we shouldn't invest time to fix it since Qt Widgets is going to be replaced eventually anyway).

**Comment by Tantacrul** [comment]
I agree. I also think you've basically designed it, so no input required from me. Let's do it!

**Comment by octopols** [comment]
Can we add a feature where the font selection drop-down also shows what the font look like?

**Comment by cbjeukendrup** [comment]
@octopols Please create a separate issue for that

**Comment by jmwmusic** [comment]
Really hope this one can be addressed as it is standard accessibility practice and makes a significant difference to productivity.  Take for example adding the swing item from the tempo palette, currently that takes 31 down arrow presses to get to!

**Comment by Tantacrul** [comment]
@jmwmusic if you can, would you avoid 'hurry up' messages, which sends us all emails? It'll get done when it gets done. The fact that it's an active issue is enough.

**Comment by jmwmusic** [comment]
Sorry, was not meaning that as a hurry up, just checking it was still being looked at as the other ticket I opened got a response added that there was no plan for this, which seemed strange as it also linked back to this one, so just making sure the need for it as an accessibility issue was understood.  I think I didn't find it in my initial search as already existing because it didn't have an accessibility tag possibly until now?

**Comment by cbjeukendrup** [comment]
The "closed as not planned" email is a standard GitHub thing, used for things like duplicates as well as things that are by design; basically, any reason apart from that the issue is solved. So in the case of a duplicate, it just means that the duplicate is moved out of the way, while the original continues to exist. 
(Don't worry about not finding this issue; sometimes finding the right search term is not quite obvious.)

**Comment by jmwmusic** [comment]
Can I check whether the scope of this issue currently includes palette lists, and if not can that be added?

**Comment by jmwmusic** [comment]
I think this might be related - Currently screen reader is not announcing the Mnemonic Shortcut options in menus. So for example in the File menu if you arrow down to Open it just says "Open Ctrl O" but would usually say "Open Ctrl O O".  But maybe this is deliberate till it is working as currently if I press Alt F O I get stuck in nomans land with no focus, tab & arrow do not work and I have to press Esc to get out of it.

**Comment by cbjeukendrup** [comment]
I think the Mnemonics issue is unrelated, and it would be preferable to have a separate issue for that. MuseScore uses a custom menu implementation on Windows, so speech for mnemonic shortcuts was probably just not implemented. However, mnemonics themselves _should_ be working correctly. If not, that would be a regression and would deserve an issue on its own.

**Comment by re1san** [comment]
Hey, I would like to work on it :)

**Comment by scorster** [comment]
Issue reported [here on .org](https://musescore.org/en/node/362510#comment-1235959), along with[issues with the Up arrow keystroke as well](https://github.com/musescore/MuseScore/issues/22179)

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/366117

**Comment by shubham-shinde-442** [comment]
Working on this.

**Comment by re1san** [comment]
@cbjeukendrup I worked on this issue earlier. It seems like the focus is disabled for dropdowns by this line: `openPolicies: PopupView.NoActivateFocus` in `StyledDropdownView.qml`. Because of this, the code in `Keys.onShortcutOverride` doesn't work and instead the global keyboard shortcut is activated as `ShortcutsController::activate` is called on pressing any key.

@Eism mentioned that this was done intentionally so that opening dropdowns wouldn't steal focus from the current window.

Hope this helps you, @shubham-shinde-442 :)

**Comment by scorster** [comment]
*@Eism mentioned that this was done intentionally so that opening dropdowns wouldn't steal focus from the current window.*

MuseScore has exhibited lots of focus issues over the years.

Why would a score need/want to respond to keystrokes made when the user has a menu open? If they expect their keystroke to interact with the score and then notice that it doesn't they can simple escape the menu.

What am I missing here?

**Comment by shoogle** [comment]
> What am I missing here?

I'm not sure, but possible reasons might be: 

1. To allow changing the dropdown value with Up and Down arrow keys without opening it first (but I think this only works in MS3 and not MS4 anyway).

2. The screen reader might treat the dropdown as a new window and say the window name when you focus/unfocus it, which could be confusing/unhelpful. 

3. Focusing might break speech in some other way, perhaps because we use our own accessibility system rather than Qt's system.

4. Focusing might break keyboard navigation somehow, perhaps because we use our own shortcut-based navigation rather than Qt navigation.

5. If you open the Playback or Appearance popup in the Properties panel, these popups are supposed to stay open while you navigate the score. Perhaps they share some code with dropdowns, and keeping them open means focus has to stay in the main window.

@shubham-shinde-442, you could try focusing the dropdown if you like. If you do, those are the things I would watch out for.

**Comment by jmwmusic** [comment]
It is great to see some first letter navigation in 4.7.  It would be very helpful if this was restored to Palettes also #14925
